src/Http/Http/src/BufferSegment.cs

```diff
@@ -0,0 +1,118 @@
+        /// The Start represents the offset into AvailableMemory where the range of "active" bytes begins. At the point when the block is leased
+        /// the Start is guaranteed to be equal to 0. The value of Start may be assigned anywhere between 0 and
+        /// AvailableMemory.Length, and must be equal to or less than End.
+        /// </summary>
+        public int Start { get; private set; }
+
+        /// <summary>
+        /// The End represents the offset into AvailableMemory where the range of "active" bytes ends. At the point when the block is leased
+        /// the End is guaranteed to be equal to Start. The value of Start may be assigned anywhere between 0 and
+        /// Buffer.Length, and must be equal to or less than End.
+        /// </summary>
+        public int End
+        {
+            get => _end;
+            set
+            {
+                Debug.Assert(value - Start <= AvailableMemory.Length);
+
+                _end = value;
+                Memory = AvailableMemory.Slice(Start, _end - Start);
+            }
+        }
+
+        /// <summary>
+        /// Reference to the next block of data when the overall "active" bytes spans multiple blocks. At the point when the block is
+        /// leased Next is guaranteed to be null. Start, End, and Next are used together in order to create a linked-list of discontiguous
+        /// working memory. The "active" memory is grown when bytes are copied in, End is increased, and Next is assigned. The "active"
+        /// memory is shrunk when bytes are consumed, Start is increased, and blocks are returned to the pool.
+        /// </summary>
+        public BufferSegment NextSegment
+        {
+            get => _next;
+            set
+            {
+                _next = value;
+                Next = value;
+            }
+        }
+
+        public void SetMemory(IMemoryOwner<byte> memoryOwner)
+        {
+            SetMemory(memoryOwner, 0, 0);
+        }
+
+        public void SetMemory(IMemoryOwner<byte> memoryOwner, int start, int end, bool readOnly = false)
+        {
+            _memoryOwner = memoryOwner;
+
+            AvailableMemory = _memoryOwner.Memory;
+
+            ReadOnly = readOnly;
+            RunningIndex = 0;
+            Start = start;
+            End = end;
+            NextSegment = null;
+        }
+
+        public void ResetMemory()
+        {
+            _memoryOwner.Dispose();
+            _memoryOwner = null;
+            AvailableMemory = default;
+        }
+
+        internal IMemoryOwner<byte> MemoryOwner => _memoryOwner;
+
+        public Memory<byte> AvailableMemory { get; private set; }
+
+        public int Length => End - Start;
+
+        /// <summary>
+        /// If true, data should not be written into the backing block after the End offset. Data between start and end should never be modified
+        /// since this would break cloning.
+        /// </summary>
+        public bool ReadOnly { get; private set; }
```

Comment:
Remove

Reply:
When we port this to System.IO.Pipelines, will we change the BufferSegment class there?

Reply:
Likely, this flag was for a feature that no longer exists.